File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/mailbox/v2/GigDetailDto.kt
@file:Suppress("PackageNaming", "ReturnCount")

package app.pantopus.android.data.api.models.mailbox.v2

import app.pantopus.android.data.api.models.common.JsonValue

/**
 * Gig-shaped sub-payload decoded from `mail.object_payload` when
 * `mail_type == "gig"` — the A17.6 "bid on your gig" mail body. Like the
 * other V2 category payloads the shape is untyped on the wire (stored as
 * JSON); the DTO is defensive and [decodeFromObjectPayload] returns null
 * unless the bidder name, bid amount, and post title are all present, so
 * the body falls back to the placeholder layout.
 */
data class GigDetailDto(
    val gigId: String? = null,
    val bidId: String? = null,
    /**
     * True once the recipient has accepted this bid — swaps the three-way
     * action row for the next-steps timeline + an "Open thread" CTA.
     */
    val isAccepted: Boolean,
    val bidder: Bidder,
    val bid: Bid,
    val post: Post,
    val otherBids: List<OtherBid>,
    val nextSteps: List<NextStep>,
) {
    /** Optimistic flip into the accepted state when the recipient accepts. */
    fun accepted(): GigDetailDto = copy(isAccepted = true)

    /** The neighbor who placed the bid. */
    data class Bidder(
        val initials: String,
        val name: String,
        val handle: String,
        val blurb: String,
        val rating: Double,
        val jobs: Int,
        val responseTime: String,
        val identityLabel: String,
        val isVerified: Boolean,
        val badges: List<String>,
    )

    /** The bid itself: amount + timing + the bidder's note. */
    data class Bid(
        val amount: Int,
        val unit: String,
        val eta: String,
        val expires: String,
        val message: List<String>,
    )

    /** Summary of the gig being bid on — tappable, opens the gig thread. */
    data class Post(
        val title: String,
        val categoryLabel: String,
        val posted: String,
        val expires: String,
        val budget: String,
        val schedule: String,
        val location: String,
        val details: String,
        val bidCount: Int,
    )

    /** A competing bid surfaced in the comparison strip. */
    data class OtherBid(
        val id: String,
        val who: String,
        val initials: String,
        val amount: Int,
        val rating: Double,
        val jobs: Int,
        val whenText: String,
        val flag: String?,
    )

    /** Lifecycle state of a next-step row in the accepted timeline. */
    enum class StepState { Active, Pending, Upcoming }

    /** One row of the post-acceptance next-steps timeline. */
    data class NextStep(
        val id: String,
        val label: String,
        val whenText: String,
        val state: StepState,
    )

    companion object {
        fun decodeFromObjectPayload(payload: JsonValue?): GigDetailDto? {
            if (payload == null) return null
            val bidderMap = payload["bidder"] as? Map<*, *>
            val bidMap = payload["bid"] as? Map<*, *>
            val postMap = payload["post"] as? Map<*, *>
            val bidder = decodeBidder(bidderMap) ?: return null
            val bid = decodeBid(bidMap) ?: return null
            val post = decodePost(postMap) ?: return null
            val gigId =
                stringCandidate(payload, "gig_id", "gigId", "gig")
                    ?: stringCandidate(postMap.orEmpty(), "id", "gig_id", "gigId")
            val bidId =
                stringCandidate(payload, "bid_id", "bidId", "bid_offer_id")
                    ?: stringCandidate(bidMap.orEmpty(), "id", "bid_id", "bidId")
            return GigDetailDto(
                gigId = gigId,
                bidId = bidId,
                isAccepted = payload["is_accepted"] as? Boolean ?: false,
                bidder = bidder,
                bid = bid,
                post = post,
                otherBids = decodeOtherBids(payload["other_bids"] as? List<*>),
                nextSteps = decodeNextSteps(payload["next_steps"] as? List<*>),
            )
        }

        private fun decodeBidder(map: Map<*, *>?): Bidder? {
            val name = (map?.get("name") as? String)?.takeIf { it.isNotEmpty() } ?: return null
            return Bidder(
                initials = map["initials"] as? String ?: initials(name),
                name = name,
                handle = map["handle"] as? String ?: "",
                blurb = map["blurb"] as? String ?: "",
                rating = (map["rating"] as? Number)?.toDouble() ?: 0.0,
                jobs = (map["jobs"] as? Number)?.toInt() ?: 0,
                responseTime = map["response_time"] as? String ?: "—",
                identityLabel = map["identity"] as? String ?: "Personal",
                isVerified = map["verified"] as? Boolean ?: false,
                badges = (map["badges"] as? List<*>)?.mapNotNull { it as? String } ?: emptyList(),
            )
        }

        private fun decodeBid(map: Map<*, *>?): Bid? {
            val amount = (map?.get("amount") as? Number)?.toInt() ?: return null
            return Bid(
                amount = amount,
                unit = map["unit"] as? String ?: "flat",
                eta = map["eta"] as? String ?: "—",
                expires = map["expires"] as? String ?: "",
                message = (map["message"] as? List<*>)?.mapNotNull { it as? String } ?: emptyList(),
            )
        }

        private fun decodePost(map: Map<*, *>?): Post? {
            val title = (map?.get("title") as? String)?.takeIf { it.isNotEmpty() } ?: return null
            return Post(
                title = title,
                categoryLabel = map["category"] as? String ?: "Gig",
                posted = map["posted"] as? String ?: "",
                expires = map["expires"] as? String ?: "",
                budget = map["budget"] as? String ?: "",
                schedule = map["schedule"] as? String ?: "",
                location = map["where"] as? String ?: "",
                details = map["details"] as? String ?: "",
                bidCount = (map["bid_count"] as? Number)?.toInt() ?: 0,
            )
        }

        private fun decodeOtherBids(list: List<*>?): List<OtherBid> =
            list.orEmpty().mapIndexedNotNull { index, raw ->
                val map = raw as? Map<*, *> ?: return@mapIndexedNotNull null
                val who = map["who"] as? String ?: return@mapIndexedNotNull null
                val amount = (map["amount"] as? Number)?.toInt() ?: return@mapIndexedNotNull null
                OtherBid(
                    id = map["id"] as? String ?: "other-$index",
                    who = who,
                    initials = map["initials"] as? String ?: initials(who),
                    amount = amount,
                    rating = (map["rating"] as? Number)?.toDouble() ?: 0.0,
                    jobs = (map["jobs"] as? Number)?.toInt() ?: 0,
                    whenText = map["when"] as? String ?: "",
                    flag = map["flag"] as? String,
                )
            }

        private fun decodeNextSteps(list: List<*>?): List<NextStep> =
            list.orEmpty().mapIndexedNotNull { index, raw ->
                val map = raw as? Map<*, *> ?: return@mapIndexedNotNull null
                val label = map["label"] as? String ?: return@mapIndexedNotNull null
                NextStep(
                    id = map["id"] as? String ?: "step-$index",
                    label = label,
                    whenText = map["when"] as? String ?: "",
                    state =
                        when (map["state"] as? String) {
                            "active" -> StepState.Active
                            "pending" -> StepState.Pending
                            else -> StepState.Upcoming
                        },
                )
            }

        private fun initials(name: String): String =
            name
                .trim()
                .split(" ")
                .take(2)
                .mapNotNull { it.firstOrNull()?.toString() }
                .joinToString("")
                .uppercase()

        private fun stringCandidate(
            payload: Map<*, *>,
            vararg keys: String,
        ): String? =
            keys.firstNotNullOfOrNull { key ->
                (payload[key] as? String)?.trim()?.takeIf { it.isNotEmpty() }
            }
    }
}

File: frontend/apps/ios/Pantopus/Core/Networking/Models/Mailbox/V2/GigDetailDTO.swift
//
//  GigDetailDTO.swift
//  Pantopus
//
//  Gig-shaped sub-payload decoded from `mail.object_payload` when
//  `mail_type == "gig"` — the A17.6 "bid on your gig" mail body. Like the
//  other V2 category payloads the shape is untyped on the wire (stored as
//  JSON), so every nested field is optional and `decode(from:)` returns nil
//  when the bare-minimum set (bidder name + bid amount + post title) is
//  absent. The body then falls back to the placeholder layout.
//

import Foundation

/// One incoming bid on a gig the recipient posted, plus the gig summary,
/// the competing bids, and the post-acceptance next-steps timeline.
public struct GigDetailDTO: Sendable, Hashable {
    public let gigId: String?
    public let bidId: String?
    /// True once the recipient has accepted this bid. Drives the secondary
    /// state: the three-way action row is swapped for the next-steps
    /// timeline + an "Open thread" CTA.
    public let isAccepted: Bool
    public let bidder: Bidder
    public let bid: Bid
    public let post: Post
    public let otherBids: [OtherBid]
    public let nextSteps: [NextStep]

    public init(
        gigId: String? = nil,
        bidId: String? = nil,
        isAccepted: Bool,
        bidder: Bidder,
        bid: Bid,
        post: Post,
        otherBids: [OtherBid],
        nextSteps: [NextStep]
    ) {
        self.gigId = gigId
        self.bidId = bidId
        self.isAccepted = isAccepted
        self.bidder = bidder
        self.bid = bid
        self.post = post
        self.otherBids = otherBids
        self.nextSteps = nextSteps
    }

    /// Returns a copy flagged accepted — used by the view-model to
    /// optimistically flip into the secondary state when the recipient
    /// taps Accept.
    public func accepted() -> GigDetailDTO {
        GigDetailDTO(
            gigId: gigId,
            bidId: bidId,
            isAccepted: true,
            bidder: bidder,
            bid: bid,
            post: post,
            otherBids: otherBids,
            nextSteps: nextSteps
        )
    }

    // MARK: - Nested shapes

    /// The neighbor who placed the bid.
    public struct Bidder: Sendable, Hashable {
        public let initials: String
        public let name: String
        public let handle: String
        public let blurb: String
        public let rating: Double
        public let jobs: Int
        public let responseTime: String
        /// Identity-pillar chip label (e.g. "Personal").
        public let identityLabel: String
        public let isVerified: Bool
        public let badges: [String]

        public init(
            initials: String,
            name: String,
            handle: String,
            blurb: String,
            rating: Double,
            jobs: Int,
            responseTime: String,
            identityLabel: String,
            isVerified: Bool,
            badges: [String]
        ) {
            self.initials = initials
            self.name = name
            self.handle = handle
            self.blurb = blurb
            self.rating = rating
            self.jobs = jobs
            self.responseTime = responseTime
            self.identityLabel = identityLabel
            self.isVerified = isVerified
            self.badges = badges
        }
    }

    /// The bid itself: amount + timing + the bidder's note.
    public struct Bid: Sendable, Hashable {
        public let amount: Int
        public let unit: String
        public let eta: String
        public let expires: String
        public let message: [String]

        public init(amount: Int, unit: String, eta: String, expires: String, message: [String]) {
            self.amount = amount
            self.unit = unit
            self.eta = eta
            self.expires = expires
            self.message = message
        }
    }

    /// Summary of the gig being bid on — tappable, opens the gig thread.
    public struct Post: Sendable, Hashable {
        public let title: String
        public let categoryLabel: String
        public let posted: String
        public let expires: String
        public let budget: String
        public let schedule: String
        public let location: String
        public let details: String
        public let bidCount: Int

        public init(
            title: String,
            categoryLabel: String,
            posted: String,
            expires: String,
            budget: String,
            schedule: String,
            location: String,
            details: String,
            bidCount: Int
        ) {
            self.title = title
            self.categoryLabel = categoryLabel
            self.posted = posted
            self.expires = expires
            self.budget = budget
            self.schedule = schedule
            self.location = location
            self.details = details
            self.bidCount = bidCount
        }
    }

    /// A competing bid surfaced in the comparison strip.
    public struct OtherBid: Sendable, Hashable, Identifiable {
        public let id: String
        public let who: String
        public let initials: String
        public let amount: Int
        public let rating: Double
        public let jobs: Int
        public let whenText: String
        /// "cheapest" / "top-rated" flag, or nil.
        public let flag: String?

        public init(
            id: String,
            who: String,
            initials: String,
            amount: Int,
            rating: Double,
            jobs: Int,
            whenText: String,
            flag: String?
        ) {
            self.id = id
            self.who = who
            self.initials = initials
            self.amount = amount
            self.rating = rating
            self.jobs = jobs
            self.whenText = whenText
            self.flag = flag
        }
    }

    /// Lifecycle state of a next-step row in the accepted timeline.
    public enum StepState: String, Sendable, Hashable {
        case active, pending, upcoming
    }

    /// One row of the post-acceptance next-steps timeline.
    public struct NextStep: Sendable, Hashable, Identifiable {
        public let id: String
        public let label: String
        public let whenText: String
        public let state: StepState

        public init(id: String, label: String, whenText: String, state: StepState) {
            self.id = id
            self.label = label
            self.whenText = whenText
            self.state = state
        }
    }

    // MARK: - Decode

    /// Best-effort decode from a JSON envelope. Returns nil unless the
    /// bidder name, bid amount, and post title are all present.
    public static func decode(from value: JSONValue?) -> GigDetailDTO? {
        guard let dict = value?.dictValue,
              let bidderDict = dict["bidder"]?.dictValue,
              let bidDict = dict["bid"]?.dictValue,
              let postDict = dict["post"]?.dictValue else { return nil }

        guard let name = bidderDict["name"]?.stringValue, !name.isEmpty,
              let amount = bidDict["amount"]?.numberValue,
              let title = postDict["title"]?.stringValue, !title.isEmpty else { return nil }

        let bidder = Bidder(
            initials: bidderDict["initials"]?.stringValue ?? Self.initials(from: name),
            name: name,
            handle: bidderDict["handle"]?.stringValue ?? "",
            blurb: bidderDict["blurb"]?.stringValue ?? "",
            rating: bidderDict["rating"]?.numberValue ?? 0,
            jobs: Int(bidderDict["jobs"]?.numberValue ?? 0),
            responseTime: bidderDict["response_time"]?.stringValue ?? "—",
            identityLabel: bidderDict["identity"]?.stringValue ?? "Personal",
            isVerified: bidderDict["verified"]?.boolValue ?? false,
            badges: (bidderDict["badges"]?.arrayValue ?? []).compactMap(\.stringValue)
        )

        let bid = Bid(
            amount: Int(amount),
            unit: bidDict["unit"]?.stringValue ?? "flat",
            eta: bidDict["eta"]?.stringValue ?? "—",
            expires: bidDict["expires"]?.stringValue ?? "",
            message: (bidDict["message"]?.arrayValue ?? []).compactMap(\.stringValue)
        )

        let post = Post(
            title: title,
            categoryLabel: postDict["category"]?.stringValue ?? "Gig",
            posted: postDict["posted"]?.stringValue ?? "",
            expires: postDict["expires"]?.stringValue ?? "",
            budget: postDict["budget"]?.stringValue ?? "",
            schedule: postDict["schedule"]?.stringValue ?? "",
            location: postDict["where"]?.stringValue ?? "",
            details: postDict["details"]?.stringValue ?? "",
            bidCount: Int(postDict["bid_count"]?.numberValue ?? 0)
        )

        let otherBids: [OtherBid] = (dict["other_bids"]?.arrayValue ?? []).enumerated().compactMap { index, raw in
            guard let bidDict = raw.dictValue,
                  let who = bidDict["who"]?.stringValue,
                  let amount = bidDict["amount"]?.numberValue else { return nil }
            return OtherBid(
                id: bidDict["id"]?.stringValue ?? "other-\(index)",
                who: who,
                initials: bidDict["initials"]?.stringValue ?? Self.initials(from: who),
                amount: Int(amount),
                rating: bidDict["rating"]?.numberValue ?? 0,
                jobs: Int(bidDict["jobs"]?.numberValue ?? 0),
                whenText: bidDict["when"]?.stringValue ?? "",
                flag: bidDict["flag"]?.stringValue
            )
        }

        let nextSteps: [NextStep] = (dict["next_steps"]?.arrayValue ?? []).enumerated().compactMap { index, raw in
            guard let stepDict = raw.dictValue,
                  let label = stepDict["label"]?.stringValue else { return nil }
            return NextStep(
                id: stepDict["id"]?.stringValue ?? "step-\(index)",
                label: label,
                whenText: stepDict["when"]?.stringValue ?? "",
                state: StepState(rawValue: stepDict["state"]?.stringValue ?? "upcoming") ?? .upcoming
            )
        }

        return GigDetailDTO(
            gigId: Self.stringCandidate(in: dict, keys: ["gig_id", "gigId", "gig"])
                ?? Self.stringCandidate(in: postDict, keys: ["id", "gig_id", "gigId"]),
            bidId: Self.stringCandidate(in: dict, keys: ["bid_id", "bidId", "bid_offer_id"])
                ?? Self.stringCandidate(in: bidDict, keys: ["id", "bid_id", "bidId"]),
            isAccepted: dict["is_accepted"]?.boolValue ?? false,
            bidder: bidder,
            bid: bid,
            post: post,
            otherBids: otherBids,
            nextSteps: nextSteps
        )
    }

    private static func initials(from name: String) -> String {
        let parts = name.split(separator: " ").prefix(2)
        return parts.compactMap { $0.first.map(String.init) }.joined().uppercased()
    }

    private static func stringCandidate(in dict: [String: JSONValue], keys: [String]) -> String? {
        for key in keys {
            guard let value = dict[key]?.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines),
                  !value.isEmpty else { continue }
            return value
        }
        return nil
    }
}
